### PR TITLE
Segregated [change layout]

### PR DESCRIPTION
**Pull request by matkoniecz**
If desirable I may split ImageList modification to a separate PR.

### CODE REVIEW

File: app/src/main/java/de/westnordost/streetcomplete/quests/ImageListQuestAnswerFragment.java [hunk 1]
```diff
@@ -85,7 +85,7 @@
 
 	protected int getItemsPerRow()
 	{
-		return 4;
+		return Math.min(4, getItems().length);
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/ImageListQuestAnswerFragment.java]: This and the changes to the other quests is not a good idea. The quest types below have 3 per row, not because they are exactly three choices, but the size of photo-selection-items should be not still smaller than that. Pictograms and icons are usually clear enough that they can be smaller.

File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java [hunk 2]
```diff
@@ -121,6 +122,7 @@
 				new AddWheelChairAccessToilets(o),
 				new AddReligionToWaysideShrine(o),
 				new AddBikeParkingType(o),
+				new AddSegregated(o),
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java]: I think what is segregated could be put in the class name here

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	@Override
+	protected String getTagFilters() {
+		return
+			"ways with " +
+				"(" +
+				"(highway = path and bicycle = designated and foot = designated)" +
+				" or (highway = footway and bicycle = designated)" +
+				" or (highway = cycleway and foot = designated)" +
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: Why only designated?

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+				")" +
+				" and surface !~" + TextUtils.join("|", OsmTaggings.ANYTHING_UNPAVED) +
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: or rather `surface ~ ...ANYTHING_PAVED`?

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	@Override
+	public AbstractQuestAnswerFragment createForm() {
+		return new AddSegregatedForm();
+	}
+
+	@Override
+	public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes) {
+		List<String> values = answer.getStringArrayList(AddPowerPolesMaterialForm.OSM_VALUES);
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: PowerPolesMaterialForm?

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -609,5 +609,8 @@ Otherwise, you can download another keyboard in the app store. Popular keyboards
     <string name="quest_address_answer_no_housenumber_message2">Is this correct and is it not just part of a building?</string>
 
     <string name="notification_channel_download">"Download"</string>
+    <string name="quest_segregated_title">How footway and cycleway are combined here?</string>
```
Comment (westnordost) [inline app/src/main/res/values/strings.xml]: Hmm, my first suggestion is:

How are the foot- and cycleway laid out here?

1. Segregated from one another
2. Cyclists and pedestrians share the space

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 3 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: because (at least in my experience) highway=footway + bicycle=yes is always segregated=no

Is it possible that bicycle has dedicated part of path and the way is not dedicated to usage by cyclists?

File: app/src/main/java/de/westnordost/streetcomplete/quests/ImageListQuestAnswerFragment.java (hunk 1 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/ImageListQuestAnswerFragment.java]: OK, I will drop this part.

> Pictograms and icons are usually clear enough that they can be smaller.

Is it still desirable to show two small icons with blank space on the right side (that is what happens where there is not enough entries to fill even the first row?

Generally, I see not even single quest where getItemsPerRow was set to value higher than count of possible answers.

**Comment by matkoniecz** [comment]
> I will draw the pictograms and an icon

Thanks!

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 5 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: ops

File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java (hunk 2 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java]: Previously it was baroque AddSegregatedStatusOfCombinedFootwayAndCycleway but it was ridiculous. 

AddIsCyclewaySegregated?

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 3 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: I can change that but I am curious how correctly tagged "highway=footway + bicycle=yes + segregated=yes" or "highway=cycleway + foot=yes + segregated=yes" looks like.

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 4 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: This quest is after surface quest so either 

- surface is added (so unpaved should exclude, at least once #746 is fixed)
- user disabled surface quest

So the difference exists when

- user added unpaved, until #746 is fixed (not showing is preferable)
- surface is unpaved, user disabled surface quest (not showing is preferable)
- surface is paved, user disabled surface quest (showing is preferable)

So in the end I planned to convince you that `surface !~ ANYTHING_UNPAVED` is preferable but rather convinced myself.

File: app/src/main/res/values/strings.xml (hunk 6 above)
Comment (matkoniecz) [reply app/src/main/res/values/strings.xml]: Thanks, I really like this version.

File: app/src/main/java/de/westnordost/streetcomplete/quests/ImageListQuestAnswerFragment.java (hunk 1 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/ImageListQuestAnswerFragment.java]: removed from this PR

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 4 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: I will take list of paved surfaces from https://github.com/gravitystorm/openstreetmap-carto/blob/ace127056386dc7aa0dd0b9e428d049c986e5df7/project.mml#L500

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 3 above)
Comment (rugk) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: Wait… we indeed [discussed that already](https://github.com/westnordost/StreetComplete/issues/527#issuecomment-324309986).
The sign, @ENT8R linked or included there – which link is apparently gone – is this one:
grafik

Just had to re-read what I've written there and seems I got to this, in the end:
> TL;DR: Quest should also ask for foot/bicycle=yes values as a legal designation via street sign […] is not needed for segregation. And pictures should be shown instead of street signs for the same reason.

So, keep it simple: Is it segregated or not?

Comment (rugk) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: AFAIK the issue was that `designated` implies a legal meaning, and, as such, `yes` can be used if there is no explicit sign.

**The images you look for were posted here**: https://github.com/westnordost/StreetComplete/issues/527#issuecomment-325822251

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	public AddSegregated(OverpassMapDataDao overpassServer) {
+		super(overpassServer);
+	}
+
+	@Override
+	protected String getTagFilters() {
+		return
+			"ways with " +
+				"(" +
+				"(highway = path and bicycle = designated and foot = designated)" +
```
[the remark this answers is not in the document]
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: > BTW, did you consider what happens if this quest get's tagging invalid values? Such as highway = footway and foot = no (and e.g. bicycle=designated to trigger this quest)?

No, but I will do it now. A good point, thanks for catching.

------

Such combination appears 25 times worldwide ( http://overpass-turbo.eu/s/Aj7 ) so I will happily ignore it.

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 3 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: >  The sign, @ENT8R linked or included there – which link is apparently gone – is this one:

Yes, I know about such places. But every single one that I encountered with this sign of equivalent is not segregated. That is why I asked for a photo of such location.

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rugk) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: Of course it is rare – everything else would be embarrassing. :wink: 
But I don't think ignoring is the best way here, because we are not talking about including tags/quests, because they are rare, but about a mistagging that may affect this quest.

---

Thinking about it, however, I guess it really is not bad to show this quest for such situations. Because it does not worsen the mistagged information (i.e. now the footway/foot=no is tagged with segregated=no or so, well… who cares?) and one also does not expect that SC would fix these issues.

----

Generally maybe it would be a good idea to add these two _other answers_:
> "This is a cycleway only" | "This is a footway only"

Just leaving a note for these small changes, which are very easy to identify for a surveyor looking at the path, would be a bit too much, IMHO.

File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java (hunk 2 above)
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java]: AddCyclewaySegregated?

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 3 above)
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: To add my two cents: 
At least in Germany, and apparently documented that way in the German wiki (didn't check), there exists the opinion that only cycleways which are compulsory to use should be tagged with cycleway=designated. And I think this has been documented like this already a long time because I remember when I started with OSM, I decided to rather not tag any cycleway as `designated` (from satellite imagery) but rather as just `yes` because I was not sure if it was compulsory to use. See also this forum topic https://forum.openstreetmap.org/viewtopic.php?id=62974 to see that this opinion is still around.
Long story short, and wrong as this "only in Germany, `designated` has a special additional meaning" may be, it has been practised like this for a long time, so chances are that a lot of these situations are around.

If the solution here should be to include all `yes` as well then, ... well, this is something we should discuss.

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: Regarding your last suggestion, @rugk , if you look outside "signposted" Germany, you will find that many sidewalks and paths will have no signs at all. How would one distinguish the case that pedestrians and cyclists are mixed from that no cyclists are allowed (i.e. because surveyor doesn't see one right now or subjectively thinks cyclists should not use this path)?

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 3 above)
Comment (rugk) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: I guess it also never does any harm to include `yes`. These are only a few extra roads then, ….

File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java (hunk 2 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java]: AddCyclewaySegregated sounds like it is about mapping cycleways. AddCyclewaySegregatedStatus?

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 3 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: "only a few extra roads" - my city has 970+ highway=footway + bicycle=yes without segregated, all of them are segreagted=no ( http://overpass-turbo.eu/s/Ajo )

File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java (hunk 2 above)
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java]: AddCyclewaySegregation?

Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.java]: that is good, thanks

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 3 above)
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: Well, then perhaps not include footways with `cycleway=yes` after all. The Germans just need to get their act together and fix their tagging practice.

File: app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rugk) [reply app/src/main/java/de/westnordost/streetcomplete/quests/segregated/AddSegregated.java]: Oh, yeah, indeed difficult. So rather leave them out.

**Comment by westnordost** [comment]
The drawables are there now, btw

**Comment by rugk** [comment]
So another thing: What about including side lanes? There, those features are likely most relevant.

**Comment by matkoniecz** [comment]
> What about including side lanes?

Can you clarify what you mean? It is unclear what and how is supposed to be included.

**Comment by rugk** [comment]
Sorry, meant sidewalks. They may also be queried and the same tags (with prefix) can be added. Example: https://github.com/westnordost/StreetComplete/issues/527#issuecomment-326985701

**Comment by matkoniecz** [comment]
It is about "#527 (comment) but not mapped as a separate path."? I see no good reason to handle it in the same quest - it requires a completely different query and applied tags are also different.

**Comment by rugk** [comment]
Yeah it's about the comment I've linked (it's the beginning).

E.g.:
```
cycleway:right=track
cycleway:right:bicycle=designated
cycleway:right:segregated=yes
```

Or with sidewalk:
```
sidewalk=right sidewalk:right:foot=designated sidewalk:right:bicycle=yes sidewalk:right:oneway=*
```

>  I see no good reason to handle it in the same quest 

Excect taht it is totally the ame question, UI and everything. Okay, maybe slight changes to indicate the user what way the quest asks about, but it's the same topic with the same possible quest.

> applied tags are also different.

Not really, you just have to prefix them. AFAIK this was also done in other quests like this.

**Comment by matkoniecz** [comment]
It requires different UI (to communicate and get left/right info), a completely different query and requires handling of prefixed for added tags.

Only quest icon will be the same.

**Comment by rugk** [comment]
I thought it could just ask for the left and right sidewalk separately, then the UI is the same. In a minimal way, the question alone could mention, which sidewalk it is asked about.

**Comment by matkoniecz** [comment]
You need to communicate to user which sidewalk is right and left (it would probably reuse cycleway/oneway quest UI)

**Comment by matkoniecz** [comment]
I am unsure is it important enough to block merging or is it a pointless bikeshedding.

At least on my phone, the initial quest UI view shows almost entire selection area, but a small part is obscured and needs to be pulled up. I think that it would be preferable to show all of it (similar things happens in some other quests).

The second issue is that the text on the "shared space" symbol is on part of pedestrian symbol what is not looking ideal.


screenshot_2018-07-23-13-18-22-906_de westnordost streetcomplete debug

**Comment by westnordost** [comment]
Hmm right, this does not look nice. You can easily change the size of the xml vector icons by modifying the android:width/height property.

Quest forms that show icons usually use another item layout which shows the text below the icon, not above it. See for example the sports quests. Also possible would be to show the text right of the icon, like in the building quest (but in two columns, if there is space).

**Comment by westnordost** [comment]
done and tested.

device-2018-08-15-215158

**Comment by matkoniecz** [comment]
Sorry, I had no time recently - and thanks for finishing this!

**Comment by rugk** [comment]
BTW while adapting the overpass-query for overpass-turbo for the Quests wiki, I noticed that this quest seems to be _extremely_ rare. In whole Rom (in Den Haag: only 8  and Netherlands are said to be bicycle-friendly) only one element was returned. When removing the "paved street" limitation it were a few more, at least.
I guess, what limits this quest, is not only this "paved street", but even more the strong "designated" requirements. Maybe they are not always tagged correctly or so.

So just FYI, maybe the data is also already so good that most places have `segregated` already. 😉

**Comment by westnordost** [comment]
I rather think this is because most of these ways are actually tagged something with ...=yes rather than designated. But well, I think we already discussed that the quest should only show on designated.

Though, perhaps it could also show on cycleway +foot=yes.

**Comment by westnordost** [comment]
Also, another reason why few of it shows is that it is only shown for paved - so if the surface hasn't been tagged yet, the quest also doesn't appear, which is okay

**Comment by rugk** [comment]
> Though, perhaps it could also show on cycleway +foot=yes.

I'd also say so, but OP said:

> because (at least in my experience) highway=footway + bicycle=yes is always segregated=no

Which may be true, but I guess many things are not correctly tagged like this and pay attention to this difference. Maybe that's why the wiki says, "segregated" should always be tagged.

In any case, given the low results of the quest, I feel it would not hurt to decrease the limits anymore. An explicitly tagged "segregated" very certainly does not hurt anyone, but it may help, as it may not always be clear that "highway=footway + bicycle=yes" (or reverse) is this "Footway, but 🚲 are allowed".

**Comment by westnordost** [comment]
I meant `highway=cycleway + foot=yes`, not `highway=footway + bicycle=yes`.

**Comment by rugk** [comment]
Ah, I get your point. Yes, it really makes no sense, or, at least, I have not seen it anywhere, yet.
